velox/docs/develop/timestamp.rst

```diff
@@ -0,0 +1,301 @@
+=================================
+Timestamp and Timezone Management
+=================================
+
+Concepts
+--------
+
+Following ANSI SQL semantics, TIMESTAMP is a data type that represents a
+reading of a wall clock and a calendar, e.g, ``2024-04-09 18:25:00``. Note that
+a TIMESTAMP does not represent an absolute point in time, as the exact same
+wall clock time may be read in different instants in time depending on where
+one is situated on earth. For example, ``2024-04-09 18:25:00`` in California
+and in China were perceived at different absolute points in time, about 15h
+apart.  
+
+To represent absolute points in time, SQL defines a TIMESTAMP WITH TIMEZONE
+type, which conceptually represents a pair of a wall time and calendar read
+(say, ``2024-04-09 18:25:00``), and a timezone (``PST``, or
+``America/Los_Angeles``). With these two values, one can unambiguously
+represent an absolute instant in time. 
+
+Naturally, a TIMESTAMP WITH TIMEZONE can be cast into a TIMESTAMP by just
+ignoring the timezone and keeping the timestamp wall time, and a TIMESTAMP can
+be cast into a TIMESTAMP WITH TIMEZONE by associating a timezone to it. The
+timezone can either be explicitly specified by the users, or implicitly taken
+from the user system or session information. 
+
+Physical Representation
+-----------------------
+
+Representing timestamps in memory as a string or a set of values for year,
+month, day, hour, and so on, is inefficient. Therefore, timestamps are usually
+stored in a columnar layout as a 64 bit integer representing the number of
+seconds elapsed since ``1970-01-01 00:00:00``. Negative values represent time
+prior to that. 
+
+However, note that the physical representation of the timestamp is orthogonal
+to its logical meaning. For example, the timestamp represented by the ``0``
+integer was perceived at different absolute points in time depending on the
+observer’s timezone, and does not necessarily imply that it was observed in the
+UTC timezone. When a timestamp represents the number of seconds in UTC
+specifically (at that exact absolute instant in time), it may be called a *unix
+epoch* or *unix time*.
+
+Velox Classes and APIs
+----------------------
+
+Velox provides a few classes and APIs to allow developers to store, process,
+and convert timestamps across timezones:
+
+**Timestamp:** In Velox, timestamps are represented by the `Timestamp
+<https://github.com/facebookincubator/velox/blob/main/velox/type/Timestamp.h>`_
+class. The Timestamp class stores two 64 bit integers, one containing the
+number of seconds from ``1970-01-01 00:00:00``, and one containing the
+nanoseconds offset in that particular second, in order to provide nanosecond
+precision. A few more observations:
+
+* While “seconds” can be negative to represent time before ``1970-01-01
+  00:00:00``, “nanoseconds” are always positive.
+
+* Engines like Presto and Spark only support milliseconds precision, so
+  naturally “nanoseconds” are zeroed.
```

Comment:
Nanoseconds are used even with millisecond precision.

Reply:
Rephrased it